ota/README.md

```diff
@@ -0,0 +1,69 @@
+# Gateway OTA updater.
+
+This folder contains the package which deals with OTA updates for the gateway. The OTA process includes downloading, extracting and installing : node, build assets, native dependencies and `node_modules`. OTA does not update any of the underlying system libraries.
+
+## Design
+
+The OTA updater is based around a "write ahead log" which keeps track of the current running state of the system, updates, downloads, extraction and installation. The OTA updater is designed around a A/B system. When A is running A will download updates to B and if they apply successfully B is started. If B somehow fails or crashes prior to successfully launching the gateway B will fall back to A after a fixed number of retries (TBD). A/B is used for illustration the underlying design could easily have A/B/C/D/etc.. updates.
+
+### Terms:
+
+Update channel: An update "channel" is simply a base URL in which the `update.json` exists. A channel most likely (TBD) is simply a S3 bucket with a `update.json` in the root.
+
+A/B update: A system with at least one viable version of the software and another which is updated.
+Update Source: Initially the system will poll a `update.json` file from a fixed url. That file is periodically updated with information on the latest update for that update channel which the gateway is pointed.
+
+Write ahead log (WAL): Must use "atomic" append only writes. TODO: find link describing how linux kernel does this.
+
+### `update.json`.
+
+This is used for high level illustration we may need additional fields.
+
+`version`: A monotonically increasing number. We use a number rather than a UUID to allow for the underlying update channel to be "eventually consistent" where we may sometimes find a number lower than the current update.
+
+`contentAddressabeURL`: The URL which points to the new update in the format of `<hash of tar>.tar.gz`. By using a hash to address the URL it makes it more difficult to pull off a man in the middle attack (which is still possible even over SSL).
+
+```json
+{
+  "version": 1,
+  "contentAddressableURL": "5dde896887f6754c9b15bfe3a441ae4806df2fde94001311e08bf110622e0bbe.tar.gz"
+}
+```
+
+### High level operations
+
+#### Downloading
+
+`currentVersion`: The currently installed version running on the gateway. Note this version may or may not be the same as `package.json.version`.
```

Comment:
Excellent question. My use of "currentVersion" is misleading what we really want here I think is "timestamp". For example I would like to release at least once a day and that (I believe) is very inconvenient to do a full version bump. So we will have something like `package.json.version` 0.3 and in this json file we will have "timestamp" (used to figure out when to download an update) and thinking further I also want to add a "revision" field so we can tell exactly what version of the code is deployed.